File: web/src/week.ts
/**
 * Read `week.bin`, the per-lot x half-hour-of-week climatology table, and
 * blend it with a live reading to answer questions past `grid.bin`'s
 * +120 min horizon.
 *
 * The binary layout is fixed by `src/parkcast/artifacts.py`'s
 * `WEEK_HEADER_FORMAT = "<4sBIHHBI"` -- little-endian, no alignment padding --
 * so the offsets below are byte-identical to little-endian `DataView` reads.
 * This module mirrors `./artifacts.ts`'s `parseGrid` discipline throughout:
 * validate the header against the body it came with, and throw rather than
 * hand back something half-trusted.
 */
import type { WeekTable } from "./types";

/** `magic(4) version(1) builtTs(4) nLots(2) nBuckets(2) bucketMin(1) rosterId(4)`. */
export const HEADER_SIZE = 18;

/** The schema version this client knows how to read. Shared with `grid.bin`. */
export const VERSION = 1;

const MAGIC = "PCW1";

/**
 * Buckets covering a full week at `BUCKET_MIN`-minute width: `7 * 24 * 60 / 30`,
 * matching `config.WEEK_BUCKETS`. `weekBucket` and `probabilityAt` hard-code
 * this shape rather than reading it back out of a parsed `WeekTable` on every
 * call -- `parseWeek` is what makes hard-coding it safe, by refusing to parse
 * a header that declares anything else.
 */
const N_BUCKETS = (7 * 24 * 60) / 30;

/** Minutes per bucket, matching `config.CLIMATOLOGY_BUCKET_MIN`. */
const BUCKET_MIN = 30;

/**
 * No observation for this lot in this bucket.
 *
 * Emphatically *not* zero, exactly as `./artifacts.ts`'s `UNKNOWN` is not zero
 * for the grid: zero is a claim -- "reliably full at this hour" -- and this is
 * the absence of one. `probabilityAt` returns `null` here and never a number.
 */
export const WEEK_UNKNOWN = 255;

/**
 * How fast a live reading's influence decays toward climatology as the
 * horizon grows, in minutes: the reading's weight halves every this many
 * minutes. Matches `config.BLEND_HALF_LIFE_MIN`, which `Blend.predict` in
 * `src/parkcast/forecast.py` uses for the same formula server-side.
 */
export const BLEND_HALF_LIFE_MIN = 30;

/**
 * Parse `week.bin`, validating that the header describes the body it came
 * with -- see `parseGrid` in `./artifacts.ts` for why this matters: a
 * truncated download and a full one are otherwise indistinguishable, and a
 * half-trusted table indexed against the wrong roster would attach every
 * lot's forecast to the wrong lot, silently and plausibly.
 */
export function parseWeek(buf: ArrayBuffer): WeekTable {
  if (buf.byteLength < HEADER_SIZE) {
    throw new Error(`week.bin is ${buf.byteLength} bytes, shorter than its ${HEADER_SIZE}-byte header`);
  }
  const dv = new DataView(buf);
  const magic = new TextDecoder("ascii").decode(new Uint8Array(buf, 0, 4));
  if (magic !== MAGIC) {
    throw new Error(`week.bin has magic ${JSON.stringify(magic)}, expected ${MAGIC}`);
  }
  const version = dv.getUint8(4);
  if (version !== VERSION) {
    throw new Error(`week.bin is schema version ${version}, this client reads ${VERSION}`);
  }

  const builtTs = dv.getUint32(5, true);
  const nLots = dv.getUint16(9, true);
  const nBuckets = dv.getUint16(11, true);
  const bucketMin = dv.getUint8(13);
  // roster_id is an unsigned 32-bit value; getUint32 (not a bitwise op, which
  // would return it signed) keeps it that way.
  const rosterId = dv.getUint32(14, true);

  // weekBucket and probabilityAt both hard-code this shape (N_BUCKETS above);
  // a header declaring a different one would silently misalign every lookup
  // rather than fail loudly, so it is refused here instead.
  if (nBuckets !== N_BUCKETS || bucketMin !== BUCKET_MIN) {
    throw new Error(
      `week.bin declares ${nBuckets} buckets at ${bucketMin} min, this client reads ${N_BUCKETS} at ${BUCKET_MIN} min`,
    );
  }

  const expected = nLots * N_BUCKETS * 2;
  const body = buf.byteLength - HEADER_SIZE;
  if (body !== expected) {
    throw new Error(
      `week.bin body is ${body} bytes, header declares ${nLots} lots x ${N_BUCKETS} buckets x 2 = ${expected}`,
    );
  }

  return {
    nLots,
    rosterId,
    builtTs,
    cells: new Uint8Array(buf, HEADER_SIZE, expected),
  };
}

/**
 * Index of the Taipei time-of-week bucket containing `ts`, 0..335.
 *
 * Must agree exactly with `forecast.week_bucket` in `src/parkcast/forecast.py`:
 *
 *     local_min = (ts + 8*3600) // 60
 *     bucket    = (local_min // 30) % 336
 *
 * That function anchors on the bare Unix epoch and does no calendar
 * arithmetic. 1970-01-01 was a **Thursday**, so **bucket 0 is Thursday 00:00
 * Taipei, not Monday** -- an earlier draft of this plan asserted a
 * Monday-anchored table, which would have disagreed with Python by 192
 * buckets (four days) while still passing a test written to match it.
 *
 * Python's `//` floors, including for negative operands (needed because
 * bucket 0's own representative timestamp, ~-28800, is negative); JS's
 * `Math.trunc` / `|0` truncate toward zero instead, which gives the wrong
 * bucket for every negative `ts`. `Math.floor` is used here for exactly that
 * reason, on both divisions, with a final normalisation so the result is
 * never negative even when the intermediate `% 336` is (JS `%` keeps the
 * sign of its left operand; Python's does not).
 */
export function weekBucket(ts: number): number {
  const localMin = Math.floor((ts + 8 * 3600) / 60);
  const raw = Math.floor(localMin / BUCKET_MIN) % N_BUCKETS;
  return raw < 0 ? raw + N_BUCKETS : raw;
}

/**
 * The climatology probability and support for one lot at the time-of-week
 * `ts` falls in.
 *
 * `p` is `null` when the bucket has no observation for this lot -- see
 * `WEEK_UNKNOWN` -- and a real number (including `0`) otherwise. `support` is
 * the raw observation count behind that bucket, capped at 255 by the encoder;
 * it is `0` whether or not `p` is `null`, since an unshrunk bucket still
 * falls back through the lot and citywide tiers and can carry a probability
 * with zero of its own support.
 */
export function probabilityAt(
  table: WeekTable,
  lotIndex: number,
  ts: number,
): { p: number | null; support: number } {
  if (!Number.isInteger(lotIndex) || lotIndex < 0 || lotIndex >= table.nLots) {
    throw new RangeError(`lot index ${lotIndex} is outside 0..${table.nLots - 1}`);
  }
  const bucket = weekBucket(ts);
  const offset = (lotIndex * N_BUCKETS + bucket) * 2;
  const percent = table.cells[offset]!;
  const support = table.cells[offset + 1]!;
  return { p: percent === WEEK_UNKNOWN ? null : percent / 100, support };
}

/**
 * Blend a live reading toward climatology as the horizon grows, matching
 * `Blend.predict` in `src/parkcast/forecast.py`:
 *
 *     weight = 0.5 ** (minutesFromReading / BLEND_HALF_LIFE_MIN)
 *     p      = weight * (observedFree >= 1 ? 1 : 0) + (1 - weight) * climatologyP
 *
 * `minutesFromReading` must be measured from the shard's `base_data_ts` --
 * the reading -- never from `Date.now()`. The server's horizon is measured
 * from `history.latest_ts`, which is exactly what `base_data_ts` carries;
 * measuring from the wall clock would pass every unit test here and still be
 * wrong in the field by however long ago the last reading was.
 *
 * `observedFree === null` means no live reading exists for this lot at all
 * (as opposed to a reading of `0`, a genuinely full lot) -- there is nothing
 * for the reading term to contribute, so this falls back to climatology
 * alone, exactly as `Blend.predict` returns `far` outright when its own
 * `near` is `None`.
 */
export function blend(observedFree: number | null, climatologyP: number, minutesFromReading: number): number {
  if (observedFree === null) return climatologyP;
  const weight = Math.pow(0.5, minutesFromReading / BLEND_HALF_LIFE_MIN);
  const near = observedFree >= 1 ? 1 : 0;
  return weight * near + (1 - weight) * climatologyP;
}

File: src/parkcast/week.py
"""Compute the week-of-history climatology cells `artifacts.encode_week` packs.

`artifacts.py` is purely a serialisation layer -- it must not import
`forecast`, or the artifact format would depend on the forecasting model
instead of the other way round. This module is what sits between them: it
reads a `History`'s counts through `forecast.Climatology` (the same tier
chain -- bucket -> lot -> global -- the live forecast uses) and returns the
`(probability, support)` cells `encode_week` expects, one row per lot.

Rows here are keyed by whatever `lot_ids` holds, which for this module is the
`History`'s own namespaced store id (`taipei:TPE0001`) -- `encode_week` wants
the bare, published id instead, and converting between the two is Task 3's
job, not this module's.
"""
from collections.abc import Sequence

from parkcast import config, forecast


def _bucket_timestamp(bucket: int) -> int:
    """A Unix timestamp `ts` such that `forecast.week_bucket(ts) == bucket`.

    `week_bucket` computes `((ts + 8h) // 60 // CLIMATOLOGY_BUCKET_MIN) %
    WEEK_BUCKETS` -- it anchors on the bare Unix epoch and does no calendar
    arithmetic, so bucket 0 is Thursday 00:00 Taipei (1970-01-01 was a
    Thursday), not Monday. Reimplementing a Monday-anchored week here would
    silently disagree with it. Instead this inverts that exact arithmetic:
    setting `local_min = bucket * CLIMATOLOGY_BUCKET_MIN` lands the inner
    `// CLIMATOLOGY_BUCKET_MIN` back on `bucket` exactly, and `bucket` is
    always already in `range(WEEK_BUCKETS)` so the outer `% WEEK_BUCKETS` is a
    no-op; undoing the same +8h Taipei shift `week_bucket` applies recovers a
    real epoch timestamp. `test_week.py` asserts this round-trips for every
    bucket in the week against `week_bucket` itself, not against this
    derivation restated.
    """
    local_min = bucket * config.CLIMATOLOGY_BUCKET_MIN
    return local_min * 60 - 8 * 3600


def build_week_cells(
    history: forecast.History, lot_ids: Sequence[str]
) -> dict[str, list[tuple[float | None, int]]]:
    """One row per id in `lot_ids`: `config.WEEK_BUCKETS` `(probability,
    support)` cells in bucket order, ready for `artifacts.encode_week`.

    The probability is `Climatology.predict(lot_id, ts, horizon_min=0)` --
    `horizon_min` is passed as `0` only for clarity; `Climatology.predict`
    ignores it entirely and keys solely on `week_bucket(ts)` and the shrinkage
    chain. Using the model's own `predict` rather than re-deriving a rate here
    is what keeps this table and the live forecast from ever silently
    drifting apart: same tiers, same shrinkage, same fallback from an empty
    bucket to the lot rate to the global one.

    The support is that bucket's own raw observation count -- the `total`
    half of `history.counts.bucket[(lot_id, bucket)]`'s `[hits, total]` pair,
    read BEFORE shrinkage touches it. `counts.bucket` is a `defaultdict`, and
    `Climatology` holds a live reference to the same dict this function reads,
    so a plain `[...]` lookup here would silently create a zeroed entry for
    every bucket this function merely asked about; `.get` is used instead so
    reading the table can never mutate the counts it reads from.

    A bucket with no observations for this lot still carries a real
    probability -- the shrinkage chain falls back to the lot rate, then the
    citywide one, and that continuity is the entire reason this artifact
    exists -- but its support is honestly `0`, never quietly promoted into
    "no data" by way of a nonzero-looking number. A lot the corpus has never
    seen ANYWHERE -- i.e. `history` itself has no observations, not even
    enough for the global tier -- gets `None` in every cell instead, because
    that is what `Climatology.predict` returns when it has no basis for an
    answer at all; `None` here is never a stand-in for a real `0`.
    """
    climatology = forecast.Climatology(history)
    bucket_counts = history.counts.bucket
    cells: dict[str, list[tuple[float | None, int]]] = {}
    for lot_id in lot_ids:
        row: list[tuple[float | None, int]] = []
        for bucket in range(config.WEEK_BUCKETS):
            ts = _bucket_timestamp(bucket)
            probability = climatology.predict(lot_id, ts, horizon_min=0)
            support = bucket_counts.get((lot_id, bucket), (0, 0))[1]
            row.append((probability, support))
        cells[lot_id] = row
    return cells
